File: src/video_marketing/worker.py
import logging
import random
from pathlib import Path

from .models import fingerprint, utc_now
from .youtube import APIError

log = logging.getLogger(__name__)
MAX_ATTEMPTS = 8


def process(store, youtube, job, clock=utc_now):
    job = store.update(job["id"], state="uploading", attempts=job["attempts"] + 1)
    try:
        if not job["video_id"]:
            if job["publish_at"] <= clock() + 60:
                store.update(job["id"], state="missed", last_error="Publication deadline missed")
                return
            size, digest = fingerprint(Path(job["path"]))
            if (size, digest) != (job["size"], job["sha256"]):
                raise APIError(
                    "Video file changed since scheduling; restore it or create a new job"
                )
            if not job["session_url"]:
                url = youtube.start(job)
                # Commit the session before sending ANY video bytes.
                job = store.update(job["id"], session_url=url)
            video_id = youtube.upload(job)
            # Commit remote identity before setting publishAt, so retries cannot upload again.
            job = store.update(job["id"], video_id=video_id)
        if youtube.schedule(job, clock):
            store.update(job["id"], state="scheduled", last_error=None)
            log.info("Job %s: YouTube accepted schedule (video %s)", job["id"], job["video_id"])
        else:
            store.update(
                job["id"],
                state="missed",
                last_error="Upload completed privately, but publication deadline was missed",
            )
    except APIError as exc:
        retry = exc.retryable and job["attempts"] < MAX_ATTEMPTS
        delay = max(exc.retry_after, min(3600, 30 * 2 ** (job["attempts"] - 1)))
        store.update(
            job["id"],
            state="retry" if retry else "failed",
            last_error=str(exc),
            next_attempt=clock() + delay + random.uniform(0, 5),
        )
        log.warning("Job %s: %s", job["id"], exc)
    except (OSError, ValueError) as exc:
        # Avoid storing paths or remote payloads from unexpected underlying errors.
        store.update(
            job["id"],
            state="failed",
            last_error=f"Local file or data error ({type(exc).__name__}); check job inputs",
        )
        log.warning("Job %s: local file or data error", job["id"])


def run_once(store, youtube, clock=utc_now) -> int:
    """Caller holds worker_lock for the lifetime of its worker process."""
    jobs = store.due(clock())
    for job in jobs:
        process(store, youtube, job, clock)
    return len(jobs)

File: src/worker.rs
use crate::{
    model::*,
    store::Store,
    youtube::{RemoteError, YouTube},
};
use anyhow::Result;
use std::sync::Arc;
use tokio::sync::watch;

pub trait UploadService: Send + Sync {
    fn start(&self, j: &Job) -> impl std::future::Future<Output = Result<String>> + Send;
    fn upload(&self, j: &Job) -> impl std::future::Future<Output = Result<String>> + Send;
    fn schedule(&self, j: &Job) -> impl std::future::Future<Output = Result<bool>> + Send;
}
impl UploadService for YouTube {
    async fn start(&self, j: &Job) -> Result<String> {
        self.start(j).await
    }
    async fn upload(&self, j: &Job) -> Result<String> {
        self.upload(j).await
    }
    async fn schedule(&self, j: &Job) -> Result<bool> {
        self.schedule(j).await
    }
}
pub async fn process(store: &Store, yt: &impl UploadService, mut j: Job) -> Result<()> {
    let result: Result<Status> = async {
        if j.video_id.is_none() {
            if j.publish_at <= now() + 60.0 {
                return Ok(Status::Missed);
            }
            let path = j.path.clone();
            let (size, digest) =
                tokio::task::spawn_blocking(move || fingerprint(std::path::Path::new(&path)))
                    .await??;
            if size != j.size || digest != j.sha256 {
                anyhow::bail!(
                    "Video file changed since scheduling. Restore it or create a new job."
                );
            }
            if j.session_url.is_none() {
                let url = yt.start(&j).await?;
                store.session(&j.id, &url)?;
                j.session_url = Some(url);
            }
            let id = yt.upload(&j).await?;
            store.video(&j.id, &id)?;
            j.video_id = Some(id);
        }
        if yt.schedule(&j).await? {
            Ok(Status::Scheduled)
        } else {
            Ok(Status::Missed)
        }
    }
    .await;
    match result {
        Ok(status) => {
            let error = if status == Status::Missed {
                Some(
                    "Publication deadline missed. Uploaded video stays private unless a previous schedule was already accepted. Check Studio before retrying.",
                )
            } else {
                None
            };
            store.finish(&j.id, status, error, now())?;
        }
        Err(e) => {
            let remote = e.downcast_ref::<RemoteError>();
            let retry = remote.is_some_and(|e| e.retryable) && j.attempts < 8;
            let delay = remote
                .map_or(0.0, |e| e.delay)
                .max((30.0 * 2f64.powi((j.attempts.saturating_sub(1).min(16)) as i32)).min(3600.0));
            // Never include reqwest's URL-bearing error chains in persisted errors or logs.
            let message = if e.downcast_ref::<reqwest::Error>().is_some() {
                "Remote response could not be read".into()
            } else {
                e.to_string()
            };
            store.finish(
                &j.id,
                if retry { Status::Retry } else { Status::Failed },
                Some(&message),
                now() + delay,
            )?;
            tracing::warn!(job=%j.id,"Upload attempt did not complete");
        }
    }
    Ok(())
}
pub async fn run(
    store: Arc<Store>,
    yt: Arc<YouTube>,
    mut shutdown: watch::Receiver<bool>,
) -> Result<()> {
    store.recover()?;
    loop {
        if *shutdown.borrow() {
            return Ok(());
        }
        if yt.connected().await {
            if let Some(j) = store.claim(now())? {
                process(&store, yt.as_ref(), j).await?;
                continue;
            }
        }
        tokio::select! {_ = tokio::time::sleep(std::time::Duration::from_secs(5))=>{},_ = shutdown.changed()=>{return Ok(())}}
    }
}
